{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Welcome to Lung Cancer Center of Excellence Workshop 2023!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**This is data description and code snippets made for R users, if you prefer Python, please click here**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dev questrions:\n",
    "The mutations are not annotated with genes - where do we get that info?\n",
    "\n",
    "make all ids either patient id or tumor id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of content:\n",
    "* [General information](#General-information)\n",
    "* [How to download the data](#How-to_download-the-data)\n",
    "* [Required software and packages](#Required-software-and-packages)\n",
    "* [I am not able to install required software/packages on my computer, what should I do?](#I-am-not-able-to-install)\n",
    "\n",
    "Sections containing data description:\n",
    "* [Somatic mutation data](#Somatic-mutation-data)\n",
    "    * [Reading data into R](#Somatic-Reading)\n",
    "    * [Column description](#Somatic-Columns)\n",
    "    * [How to extract mutation coordinates](#Somatic-Coordinates)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## General information <a class=\"anchor\" id=\"General-information\"></a>\n",
    "\n",
    "The workshop is based on data generated by the TRACERx (Tracking Cancer Evolution through Therapy (Rx) project. It is a large-scale research project aimed at understanding how cancer evolves and spreads over time. The project involves collecting multiple samples of tumours and analysing them in detail to identify genetic changes that occur as the cancer progresses.\n",
    "\n",
    "Overall, 1,644 tumour regions were sampled at surgery or during follow-up from the first 421 patients with non-small cell lung cancer (NSCLC). The samples were subjected to whole exome sequencing (WES) and RNA-seq and thoroughly analysed. The results obtained from this analysis serve as the foundation for generating hypotheses during this workshop.\n",
    "\n",
    "The available data include:\n",
    "1. patient pathology and clinical data\n",
    "2. somatic mutations\n",
    "3. somatic copy number alterations\n",
    "4. expression data\n",
    "\n",
    "While rudimentary plots summarising the data are available below, we highly encourage you to have a brief look through the [collection of TRACERx papers](https://www.nature.com/collections/haffgaicaf). In particular, somatic point and copy number alterations are discussed [here](https://www.nature.com/articles/s41586-023-05783-5) and expression changes are examined [here](https://www.nature.com/articles/s41586-023-05706-4)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How to download the data <a class=\"anchor\" id=\"How-to_download-the-data\"></a>\n",
    "Please head to [zenodo](https://zenodo.org/record/7822002) and press download:\n",
    "\n",
    "![zenodo](support_illustrations/zenodo.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The archive `figurecode.zip` will be downloaded, please unzip it. It will produce the folder `figurecode`. While the folder contains a multitude of files, we believe a handful of them can go a long way in terms of hypothesis generation:\n",
    "\n",
    "* 20221109_TRACERx421_all_tumour_df.rds\n",
    "* [Somatic single point mutations and small indels](#Somatic-mutation-data): 20221123_TRACERx421_mutation_table_region.fst\n",
    "* Somatic copy number alterations: 20221109_TRACERx421_scna_table.fst\n",
    "\n",
    "> Coordinates of mutations and CNAs in all listed files are in hg19."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Required software and packages <a class=\"anchor\" id=\"Required-software-and-packages\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to load the data into R, perform minimal data wrangling and visualization, we reccomend using `dplyr`, `fst` and `ggplot2` packages freely avaible from CRAN. But please do feel free to use any packages you are used to or find helpful!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressWarnings(suppressPackageStartupMessages(library(dplyr)))\n",
    "suppressWarnings(suppressPackageStartupMessages(library(fst)))\n",
    "suppressWarnings(suppressPackageStartupMessages(library(ggplot2)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## I am not able to install required software/packages on my computer, what should I do? <a class=\"anchor\" id=\"I-am-not-able-to-install\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Insert here how to use google colab"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Somatic mutation data <a class=\"anchor\" id=\"Somatic-mutation-data\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading data into R <a class=\"anchor\" id=\"Somatic-Reading\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The code below shows how to load mutational data in the memory of your computer and to simplify some columns for your convinience. \n",
    "> **Please do use the snippet below to read your data in as we performed some modifications in order to streamline your future analysis**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read table into memory\n",
    "mut_table = read_fst('tracerx_data/figurecode/data/20221123_TRACERx421_mutation_table_region.fst')\n",
    "\n",
    "# simplify RegionID, please do not skip this step\n",
    "mut_table$is_lymphnode = grepl('LN', mut_table$RegionID)\n",
    "mut_table$RegionID = gsub('.*[.]R|.*LN', '', mut_table$RegionID)\n",
    "mut_table$RegionID = gsub('01', '1', mut_table$RegionID)\n",
    "lymph_ids = mut_table$is_lymphnode == T\n",
    "mut_table$RegionID[lymph_ids] = paste0('LN', mut_table$RegionID[lymph_ids])\n",
    "tumor_ids = mut_table$is_lymphnode == F\n",
    "mut_table$RegionID[tumor_ids] = paste0('R', mut_table$RegionID[tumor_ids])\n",
    "mut_table$RegionID = apply(mut_table, 1, \n",
    "                           function(x) paste0(x['tumour_id'], ':', \n",
    "                                              x['RegionID']))\n",
    "rm(tumor_ids, lymph_ids)\n",
    "\n",
    "                           \n",
    "# simplify mutation_id, please do not skip this step\n",
    "muts_to_simplify <- grepl(';', mut_table$mutation_id)\n",
    "mut_table[muts_to_simplify, ]$mutation_id = sapply(mut_table[muts_to_simplify, ]$mutation_id,\n",
    "                                                   function(x) sub(':', \n",
    "                                                                   paste0('_', \n",
    "                                                                          gsub('.*;', '', x),\n",
    "                                                                          ':'),\n",
    "                                                                   x))\n",
    "mut_table[muts_to_simplify, ]$mutation_id = gsub(';.*', '',\n",
    "                                                 mut_table[muts_to_simplify, ]$mutation_id)\n",
    "\n",
    "# simplify MajorCPN_SC and MinorCPN_SC, please do not skip this step\n",
    "mut_table$r_id = gsub('.*:', '', mut_table$RegionID)\n",
    "mut_table$r_id = paste0('.*', mut_table$r_id, ':')\n",
    "mut_table$MajorCPN_SC = apply(mut_table, 1, \n",
    "                              function(x) gsub(x['r_id'], '', x['MajorCPN_SC']))\n",
    "mut_table$MajorCPN_SC = gsub(';.*', '', mut_table$MajorCPN_SC)\n",
    "mut_table$MajorCPN_SC = suppressWarnings(as.integer(mut_table$MajorCPN_SC))\n",
    "mut_table$MinorCPN_SC = apply(mut_table, 1, \n",
    "                              function(x) gsub(x['r_id'], '', x['MinorCPN_SC']))\n",
    "mut_table$MinorCPN_SC = gsub(';.*', '', mut_table$MinorCPN_SC)\n",
    "mut_table$MinorCPN_SC = suppressWarnings(as.integer(mut_table$MinorCPN_SC))                                                 \n",
    "                                                    \n",
    "# re-order columns for clarity\n",
    "mut_table = mut_table[, c('tumour_id', 'RegionID', 'is_lymphnode', 'mutation_id', \n",
    "                          'var_count', 'depth', 'Is.present.region', \n",
    "                          'CCF_pyclone_SC', 'CCF_phylo_SC', 'mut_cpn_SC', \n",
    "                          'MajorCPN_SC', 'MinorCPN_SC', 'PyCloneCluster_SC',\n",
    "                          'PyCloneClonal_SC', 'cleanCluster_SC')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In case your computer can't handle such big data frame, you can download the pre-processed version [here]()."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Column description <a class=\"anchor\" id=\"Somatic-Columns\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 Ã— 15</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>tumour_id</th><th scope=col>RegionID</th><th scope=col>is_lymphnode</th><th scope=col>mutation_id</th><th scope=col>var_count</th><th scope=col>depth</th><th scope=col>Is.present.region</th><th scope=col>CCF_pyclone_SC</th><th scope=col>CCF_phylo_SC</th><th scope=col>mut_cpn_SC</th><th scope=col>MajorCPN_SC</th><th scope=col>MinorCPN_SC</th><th scope=col>PyCloneCluster_SC</th><th scope=col>PyCloneClonal_SC</th><th scope=col>cleanCluster_SC</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>CRUK0005</td><td>CRUK0005:R1</td><td>FALSE</td><td>CRUK0005:19:34291428:G</td><td>  5</td><td>159</td><td> TRUE</td><td>0.54</td><td>0.54</td><td>0.54</td><td>4</td><td>2</td><td>3</td><td>S</td><td>TRUE</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CRUK0005</td><td>CRUK0005:R1</td><td>FALSE</td><td>CRUK0005:19:9084770:C </td><td>149</td><td>498</td><td> TRUE</td><td>1.16</td><td>1.16</td><td>4.57</td><td>4</td><td>0</td><td>2</td><td>C</td><td>TRUE</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>CRUK0005</td><td>CRUK0005:R1</td><td>FALSE</td><td>CRUK0005:1:2160418:G  </td><td> 27</td><td>171</td><td> TRUE</td><td>0.96</td><td>0.96</td><td>2.37</td><td>2</td><td>1</td><td>2</td><td>C</td><td>TRUE</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>CRUK0005</td><td>CRUK0005:R1</td><td>FALSE</td><td>CRUK0005:19:40095981:G</td><td> 47</td><td>507</td><td> TRUE</td><td>1.21</td><td>1.21</td><td>1.25</td><td>2</td><td>0</td><td>2</td><td>C</td><td>TRUE</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>CRUK0005</td><td>CRUK0005:R1</td><td>FALSE</td><td>CRUK0005:19:41313703:C</td><td>  0</td><td>616</td><td>FALSE</td><td>0.00</td><td>0.00</td><td>0.00</td><td>2</td><td>0</td><td>5</td><td>S</td><td>TRUE</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>CRUK0005</td><td>CRUK0005:R1</td><td>FALSE</td><td>CRUK0005:20:31387110:T</td><td> 68</td><td>389</td><td> TRUE</td><td>0.92</td><td>0.92</td><td>2.76</td><td>3</td><td>1</td><td>2</td><td>C</td><td>TRUE</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 Ã— 15\n",
       "\\begin{tabular}{r|lllllllllllllll}\n",
       "  & tumour\\_id & RegionID & is\\_lymphnode & mutation\\_id & var\\_count & depth & Is.present.region & CCF\\_pyclone\\_SC & CCF\\_phylo\\_SC & mut\\_cpn\\_SC & MajorCPN\\_SC & MinorCPN\\_SC & PyCloneCluster\\_SC & PyCloneClonal\\_SC & cleanCluster\\_SC\\\\\n",
       "  & <chr> & <chr> & <lgl> & <chr> & <dbl> & <dbl> & <lgl> & <dbl> & <dbl> & <dbl> & <int> & <int> & <dbl> & <chr> & <lgl>\\\\\n",
       "\\hline\n",
       "\t1 & CRUK0005 & CRUK0005:R1 & FALSE & CRUK0005:19:34291428:G &   5 & 159 &  TRUE & 0.54 & 0.54 & 0.54 & 4 & 2 & 3 & S & TRUE\\\\\n",
       "\t2 & CRUK0005 & CRUK0005:R1 & FALSE & CRUK0005:19:9084770:C  & 149 & 498 &  TRUE & 1.16 & 1.16 & 4.57 & 4 & 0 & 2 & C & TRUE\\\\\n",
       "\t3 & CRUK0005 & CRUK0005:R1 & FALSE & CRUK0005:1:2160418:G   &  27 & 171 &  TRUE & 0.96 & 0.96 & 2.37 & 2 & 1 & 2 & C & TRUE\\\\\n",
       "\t4 & CRUK0005 & CRUK0005:R1 & FALSE & CRUK0005:19:40095981:G &  47 & 507 &  TRUE & 1.21 & 1.21 & 1.25 & 2 & 0 & 2 & C & TRUE\\\\\n",
       "\t5 & CRUK0005 & CRUK0005:R1 & FALSE & CRUK0005:19:41313703:C &   0 & 616 & FALSE & 0.00 & 0.00 & 0.00 & 2 & 0 & 5 & S & TRUE\\\\\n",
       "\t6 & CRUK0005 & CRUK0005:R1 & FALSE & CRUK0005:20:31387110:T &  68 & 389 &  TRUE & 0.92 & 0.92 & 2.76 & 3 & 1 & 2 & C & TRUE\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 Ã— 15\n",
       "\n",
       "| <!--/--> | tumour_id &lt;chr&gt; | RegionID &lt;chr&gt; | is_lymphnode &lt;lgl&gt; | mutation_id &lt;chr&gt; | var_count &lt;dbl&gt; | depth &lt;dbl&gt; | Is.present.region &lt;lgl&gt; | CCF_pyclone_SC &lt;dbl&gt; | CCF_phylo_SC &lt;dbl&gt; | mut_cpn_SC &lt;dbl&gt; | MajorCPN_SC &lt;int&gt; | MinorCPN_SC &lt;int&gt; | PyCloneCluster_SC &lt;dbl&gt; | PyCloneClonal_SC &lt;chr&gt; | cleanCluster_SC &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | CRUK0005 | CRUK0005:R1 | FALSE | CRUK0005:19:34291428:G |   5 | 159 |  TRUE | 0.54 | 0.54 | 0.54 | 4 | 2 | 3 | S | TRUE |\n",
       "| 2 | CRUK0005 | CRUK0005:R1 | FALSE | CRUK0005:19:9084770:C  | 149 | 498 |  TRUE | 1.16 | 1.16 | 4.57 | 4 | 0 | 2 | C | TRUE |\n",
       "| 3 | CRUK0005 | CRUK0005:R1 | FALSE | CRUK0005:1:2160418:G   |  27 | 171 |  TRUE | 0.96 | 0.96 | 2.37 | 2 | 1 | 2 | C | TRUE |\n",
       "| 4 | CRUK0005 | CRUK0005:R1 | FALSE | CRUK0005:19:40095981:G |  47 | 507 |  TRUE | 1.21 | 1.21 | 1.25 | 2 | 0 | 2 | C | TRUE |\n",
       "| 5 | CRUK0005 | CRUK0005:R1 | FALSE | CRUK0005:19:41313703:C |   0 | 616 | FALSE | 0.00 | 0.00 | 0.00 | 2 | 0 | 5 | S | TRUE |\n",
       "| 6 | CRUK0005 | CRUK0005:R1 | FALSE | CRUK0005:20:31387110:T |  68 | 389 |  TRUE | 0.92 | 0.92 | 2.76 | 3 | 1 | 2 | C | TRUE |\n",
       "\n"
      ],
      "text/plain": [
       "  tumour_id RegionID    is_lymphnode mutation_id            var_count depth\n",
       "1 CRUK0005  CRUK0005:R1 FALSE        CRUK0005:19:34291428:G   5       159  \n",
       "2 CRUK0005  CRUK0005:R1 FALSE        CRUK0005:19:9084770:C  149       498  \n",
       "3 CRUK0005  CRUK0005:R1 FALSE        CRUK0005:1:2160418:G    27       171  \n",
       "4 CRUK0005  CRUK0005:R1 FALSE        CRUK0005:19:40095981:G  47       507  \n",
       "5 CRUK0005  CRUK0005:R1 FALSE        CRUK0005:19:41313703:C   0       616  \n",
       "6 CRUK0005  CRUK0005:R1 FALSE        CRUK0005:20:31387110:T  68       389  \n",
       "  Is.present.region CCF_pyclone_SC CCF_phylo_SC mut_cpn_SC MajorCPN_SC\n",
       "1  TRUE             0.54           0.54         0.54       4          \n",
       "2  TRUE             1.16           1.16         4.57       4          \n",
       "3  TRUE             0.96           0.96         2.37       2          \n",
       "4  TRUE             1.21           1.21         1.25       2          \n",
       "5 FALSE             0.00           0.00         0.00       2          \n",
       "6  TRUE             0.92           0.92         2.76       3          \n",
       "  MinorCPN_SC PyCloneCluster_SC PyCloneClonal_SC cleanCluster_SC\n",
       "1 2           3                 S                TRUE           \n",
       "2 0           2                 C                TRUE           \n",
       "3 1           2                 C                TRUE           \n",
       "4 0           2                 C                TRUE           \n",
       "5 0           5                 S                TRUE           \n",
       "6 1           2                 C                TRUE           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(mut_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* `tumour_id` - unique tumor identifier\n",
    "* `RegionID` - a unique identifier for a region inside the tumor. As it is a multiregional data, one tumor can have multiple `RegionID` linked to it. An identifier has the following format: tumour_id:region name. For example, CRUK0005:R1 is an identifier for the region 1 extracted from tumor CRUK0005 during the surgery. If region name starts with **\"R\"** it means that it was extracted from the actual tumor, and if it starts with **\"LN\"** the extraction was performed from the lymph node.\n",
    "* `is_lymphnode` - boolean, indicating if extraction was performed from the actual tumor mass (FALSE) or from a lymph node (TRUE).\n",
    "* `mutation_id` - a unique identifier for detected somatic mutation. An identifier has the following format: tumour_id:chromosome:position:alternative base pair.\n",
    "* `var_count` - number of reads covering that genomic position with the alternative base (mutated base pair) detected at the position.\n",
    "* `depth` - total depth of sequencing (number of reads) at this position.\n",
    "* `Is.present.region` - boolean indicator (TRUE/FALSE) of mutation's presence in this region of a tumor.\n",
    "* `CCF_pyclone_SC` - cancer cell fraction (CCF) of a mutation derived from pyclone. CCF is a proportion of cancer cells that contain this mutation in the region, where 100 means that all cancer cells from the extracted region contain that mutation.\n",
    "* `CCF_phylo_SC` - cancer cell fraction (CCF) of a mutation derived from a phylogenetic tree.\n",
    "* `mut_cpn_SC ` - mutation copy number\n",
    "* `MajorCPN_SC` - copy number of the major allele\n",
    "* `MinorCPN_SC` - copy number of the minor allele\n",
    "* `PyCloneCluster_SC` - cluster id from pyclone to which mutation belongs\n",
    "* `PyCloneClonal_SC ` - clonality status of a mutation: clonal (C) or subclonal(S)\n",
    "* `cleanCluster_SC` - cluster id from pyclone to which mutation belongs, inconsistent clusters were removed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How to extract mutation coordinates <a class=\"anchor\" id=\"Somatic-Coordinates\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It may be useful for some of your analysis to have a separate data frame with the coordinates of all detected somatic mutations. The snippet below provides an example of how the coordinates can be derived from mutation ids.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 Ã— 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>tumour_id</th><th scope=col>chr</th><th scope=col>pos</th><th scope=col>ALT</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>CRUK0005</td><td>19</td><td>34291428</td><td>G</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CRUK0005</td><td>19</td><td>9084770 </td><td>C</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>CRUK0005</td><td>1 </td><td>2160418 </td><td>G</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>CRUK0005</td><td>19</td><td>40095981</td><td>G</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>CRUK0005</td><td>19</td><td>41313703</td><td>C</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>CRUK0005</td><td>20</td><td>31387110</td><td>T</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 Ã— 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & tumour\\_id & chr & pos & ALT\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & CRUK0005 & 19 & 34291428 & G\\\\\n",
       "\t2 & CRUK0005 & 19 & 9084770  & C\\\\\n",
       "\t3 & CRUK0005 & 1  & 2160418  & G\\\\\n",
       "\t4 & CRUK0005 & 19 & 40095981 & G\\\\\n",
       "\t5 & CRUK0005 & 19 & 41313703 & C\\\\\n",
       "\t6 & CRUK0005 & 20 & 31387110 & T\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 Ã— 4\n",
       "\n",
       "| <!--/--> | tumour_id &lt;chr&gt; | chr &lt;chr&gt; | pos &lt;chr&gt; | ALT &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 1 | CRUK0005 | 19 | 34291428 | G |\n",
       "| 2 | CRUK0005 | 19 | 9084770  | C |\n",
       "| 3 | CRUK0005 | 1  | 2160418  | G |\n",
       "| 4 | CRUK0005 | 19 | 40095981 | G |\n",
       "| 5 | CRUK0005 | 19 | 41313703 | C |\n",
       "| 6 | CRUK0005 | 20 | 31387110 | T |\n",
       "\n"
      ],
      "text/plain": [
       "  tumour_id chr pos      ALT\n",
       "1 CRUK0005  19  34291428 G  \n",
       "2 CRUK0005  19  9084770  C  \n",
       "3 CRUK0005  1   2160418  G  \n",
       "4 CRUK0005  19  40095981 G  \n",
       "5 CRUK0005  19  41313703 C  \n",
       "6 CRUK0005  20  31387110 T  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mutation_coords = strsplit(mut_table$mutation_id, \":\")\n",
    "mutation_coords = as.data.frame(do.call(rbind, mutation_coords))\n",
    "colnames(mutation_coords) = c('tumour_id', 'chr', 'pos', 'ALT')\n",
    "head(mutation_coords)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> IMPORTANT NOTES"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Plot overview of number of regions per tumor - or see the figure X\n",
    "* Plot overview of number of mutations per tumor and per region\n",
    "* Distributions of depth and var count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SCNA table\n",
    "TRACERx cohort consists of 421 patients, from which a total of 1644 tumor regions were sampled and subjected to whole-exome sequencing. Out of all these samples, there were only 401 tumors for which it was possible to reconstruct phylogenetic trees (only tumors where the purity was sufficient to determine copy number states in at least two regions were retained), for a total of 1428 regions.\n",
    "\n",
    "This table contains the information about SCNA detected in each tumor sample corresponding to a specific region, inferred from *WES* data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cna_table = read_fst('tracerx_data/figurecode/data/20221109_TRACERx421_scna_table.fst')\n",
    "head(cna_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The columns are the following:\n",
    "* `patient_id` : patient identifier\n",
    "* `tumour_id` : tumor identifier\n",
    "* `sample` Sample identfier. Example: CRUK0000_SU_T1-R1 stands for the sample of patient identifier CRUK0000 that was taken from regionâ€‰1 (R1) of tumourâ€‰1 (T1), which was resected during primary surgery (SU).\n",
    "* `chr` - `startpos` - `endpos`: coordinates of the copy number segment.\n",
    "* `n.het` : number of heterozygous SNPs called on the segment.\n",
    "* `cnTotal` : segment total copy number (ASCAT)\n",
    "* `nMajor` : copy number of the major allele (ASCAT) (?)\n",
    "* `nMinor` : copy number of the minor allele (ASCAT) (?)\n",
    "* `Ploidy` : ploidy value of the sample (ASCAT) \n",
    "* `ACF` : Aberrant Cell Fraction (fraction of tumor cells in the sample) (ASCAT)\n",
    "* `nAraw` : raw Allele specific copy number for the A-allele (ASCAT)\n",
    "* `nBraw` : raw Allele specific copy number for the B-allele (ASCAT)\n",
    "* `cpn_event_vs_ploidy` : copy number status with respect to the sample ploidy. It is a string whose value can be one of `c('neutral', 'gain', 'amp', 'loss', 'deep_loss')` E.g., if ploidy is 3 and cnTotal is 3, `cpn_event_vs_ploidy` will be `'neutral'`. If ploidy is 3 and cnTotal is 4, then `cpn_event_vs_ploidy` will be `'gain'`.\n",
    "* `MSAI` : Mirrored Subclonal Allelic Imbalance. Logical indicating whether the corresponding allele specific copy number event corresponds to a mirrored subclonal allelic imbalance.\n",
    "* `samp_used_to_phase` : ID of the sample whose allelic imbalance data was used to phase SNPs and estimate haplotype specific copy number profiles. ?: was this the result of the re-run with Refphase?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pathology and clinical data\n",
    "TRACERx cohort consists of 421 patients, harbouring 432 genomically independent tumors. Each line in this dataframe corresponds to one of these tumours, reporting clinical data and histological features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "tumor_df = readRDS('tracerx_data/figurecode/data/20221109_TRACERx421_all_tumour_df.rds')\n",
    "head(tumor_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The columns in the dataframe are the following:\n",
    "* `tumour_id_muttable_cruk` : ID of the tumor sample\n",
    "* `cruk_id` : ID of the tumor sample (same information as the previous column)\n",
    "* `clinical_sex`: patient sex\n",
    "* `age` : patient age\n",
    "* `ethnicity` : patient ethnicity. It is one of: `c('White- Irish', 'White- British', 'Caribbean', 'Black', 'White and Asian', 'White-European', 'White- Other', 'South American','Indian', 'Middle eastern', 'White and Black')`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ggplot(data = tumor_df,\n",
    "    mapping = aes(x = '', fill = ethnicity)) +\n",
    "    geom_bar() + coord_polar(\"y\", start=0) + theme(axis.text = element_blank()) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* `Histology_per_tumour_id_muttable` : full name of the tumor sample histological subtype. It's one of: `c('Invasive adenocarcinoma', 'LCNEC', 'Squamous cell carcinoma', 'Pleomorphic carcinoma', 'Adenosquamous carcinoma', 'Carcinosarcoma', 'combined LUAD and LCNEC', 'Large cell carcinoma', 'Collision LUAD and LUSC')`.  There are 248 Invasive Adenocarcinomas, 138 Squamous cell carcinomas and 46 other histological subtypes. The labels are distributed as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ggplot(data = tumor_df,\n",
    "    mapping = aes(x = Histology_per_tumour_id_muttable , fill = Histology_per_tumour_id_muttable)) +\n",
    "    geom_bar() + # coord_polar(\"y\", start=0) + \n",
    "\ttheme(legend.position = 'none', \n",
    "\t\taxis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 17), \n",
    "\t\ttext = element_text(size = 15)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "tumor_df %>% select(pN_stage_per_lesion, pT_stage_per_lesion) %>% unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* `histology_3` : for samples that are either `'Invasive adenocarcinoma'` or `'Squamous cell carcinoma'` this column is equal to LUAD and LUSC respectively. For all other histologies, this column contains 'Other'.\n",
    "* `LUAD_pred_subtype_with.IMA_per_tumour`. For LUAD samples, this column contains the tissue architecture growth pattern. Such patterns are inferred from diagnostic H&E slides, and they can be divided in two categories: low-grade (lepidic, mid grade, papillary and acinar) and high-grade (cribriform, micropapillary and solid).\n",
    "* `site_per_lesion` : location of the tumor. It is one of `c('Right Upper Lobe', 'Left Upper Lobe', 'Left Lower Lobe', 'Right Lower Lobe', 'Right Middle Lobe', 'Left Upper lobe')`\n",
    "* `pT_stage_per_lesion` : (?) I don't know what this stands for.\n",
    "* `pN_stage_per_lesion` : (?) I don't know what this stands for. The values are 0, 1 and 2\n",
    "* `size_pathology_per_lesion` : pathological tumor size. (?) how is it calculated? It ranges between 7 and 140\n",
    "* `vascular_invasion_per_lesion` : logical indicating the presence of lymphovascular invasion.\n",
    "* `pleural_invasion_per_lesion` : logical indicating the presence of visceral pleural invasion.\n",
    "* `smoking_status_merged` : smking status of the patient. Cna be `c('Ex-Smoker' 'Smoker' and 'Never Smoked')`\n",
    "* `cigs_perday` : number of cigarettes smoked per day. There are non integer values because some patients smoke cigars or pipes, whose amoiunt was converted to the equivalent in cigarettes counts. One cigar is assumed to be equivalent to approximately 1.5 cigarettes, and for pipes, one bowl of tobacco is equivalent to 2.5 cigarettes, following the table presented [here](http://www.smoking2.nes.scot.nhs.uk/module4/working-out-cigarette-equivalents.html) for conversions.\n",
    "* `years_smoking` : duration fof smoking in terms of years.\n",
    "* `pack_years` : number of cigarettes packs smoked per year.\n",
    "* `is.family.lung` : (?) I don't know what this stands for. Probably is a logical indicating whether there is a history of lung cancer in the family.\n",
    "* `pathologyTNM` : pathological Tumor, Node and Metastasis (TNM) stage (? check I defined it right)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "sort(unique(tumor_df$size_pathology_per_lesion))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "sort(unique(tumor_df$vascular_invasion_per_lesion))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "unique(tumor_df$cigs_perday)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
